### PR TITLE
Provider: Default tags support 

### PR DESCRIPTION
**Pull request by venkat-iblox**
- Updated provider to add default tags
- Updated IPAM resources with default tag support
- updated acceptance test with IPAM resource to test default tags

### CODE REVIEW

**Comment by mathewab** [comment]
Can you please update with the latest go-client and also run go fmt?

**Comment by venkat-iblox** [comment]
There are more changes coming to support Acc test + fixes for issues noticed in the data source. Will bring the latest client with this change

File: docs/guides/quickstart-dhcp.md [hunk 1]
```diff
@@ -30,6 +30,9 @@ terraform {
 provider "bloxone" {
   csp_url = "https://csp.infoblox.com"
   api_key = "<BloxOne API Key>"
+  default_tags = {
+	managed_by = "terraform"
+  }
```
Comment (mathewab) [inline docs/guides/quickstart-dhcp.md:35]: Since this is optional, don't include in quick start guides.

File: docs/index.md [hunk 2]
```diff
@@ -16,6 +16,10 @@ The BloxOne provider is used to interact with the resources supported by Infoblo
 provider "bloxone" {
   csp_url = "https://csp.infoblox.com"
   api_key = "<BloxOne DDI API Key>"
+  default_tags = {
+    managed_by = "terraform"
+    location   = "us-west-1"
+  }
```
Comment (mathewab) [inline docs/index.md:22]: I think we should have two blocks there. One with the basic config for csp_url and api_key. And another block that explains that default tags can be provided in the provider configuration as a way to define common tags and have this example under there.

File: docs/index.md [hunk 3]
```diff
@@ -26,3 +30,4 @@ provider "bloxone" {
 
 - `api_key` (String) API key for accessing the BloxOne API. Can also be configured by using the `BLOXONE_API_KEY` environment variable. https://docs.infoblox.com/space/BloxOneCloud/35430405/Configuring+User+API+Keys
 - `csp_url` (String) URL for BloxOne Cloud Services Portal. Can also be configured using the `BLOXONE_CSP_URL` environment variable.
+- `default_tags` (Map of String) Default global tags the client can set for all requests.
```
Comment (mathewab) [inline docs/index.md:33]:
```suggestion
- `default_tags` (Map of String) Tags to default for all resources```

File: docs/index.md (hunk 2 above)
Comment (venkat-iblox) [reply docs/index.md:22]: Instead, shall I add an `optional` comment above the default_tags section?

File: docs/guides/quickstart-dhcp.md (hunk 1 above)
Comment (venkat-iblox) [reply docs/guides/quickstart-dhcp.md:35]: Ok. I should ideally create a separate section on its usage.

File: docs/index.md (hunk 3 above)
Comment (venkat-iblox) [reply docs/index.md:33]: Will make the change in provider.go schema

File: internal/provider/provider.go [hunk 4]
```diff
@@ -1,192 +1,220 @@
...
-		APIKey:     data.APIKey.ValueString(),
-		CSPURL:     data.CSPUrl.ValueString(),
-	})
-	if err != nil {
-		resp.Diagnostics.AddError("Client error", fmt.Sprintf("Unable to create new API client: %s", err))
-	}
+    if p.defaultTags != nil {
+        dfTags = p.defaultTags
+    } else {
+        dfTags = utils.DefaultTagsHandler(flex.ExpandFrameworkMapString(ctx, data.DefaultTags, &resp.Diagnostics), &resp.Diagnostics)
```
Comment (mathewab) [inline internal/provider/provider.go:79]: This is doing a double conversion. From `map[string]string` to `map[string]interface{}` and back to `map[string]string`

You don't need the DefaultTagsHandler, and can instead do the conversion directly. 
Also, here p.defaultTags takes precedence. IMO, the configured value should still take precendence. If configured value is not given, p.defaultTags should be used. 

```
    dfTags := p.defaultTags
    if !data.DefaultTags.IsNull() && !data.DefaultTags.IsUnknown() {
        dfTags := make(map[string]string, len(data.DefaultTags.Elements()))
        resp.Diagnostics.Append(data.DefaultTags.ElementsAs(ctx, &dfTags, false)...)
    }
```

File: internal/acctest/acctest.go [hunk 5]
```diff
@@ -9,11 +9,18 @@ import (
...
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
 
 	bloxoneclient "github.com/infobloxopen/bloxone-go-client/client"
 	"github.com/infobloxopen/terraform-provider-bloxone/internal/provider"
 )
 
+const (
+	letterBytes  = "abcdefghijklmnopqrstuvwxyz"
+	defaultKey   = "managed_by"
+	defaultValue = "terraform"
```
[the remark this answers is not in the document]
Comment (venkat-iblox) [reply internal/acctest/acctest.go:21]: It's used in 2 different places, hence the constants

**Review by mathewab — APPROVED** [review]

**Pull request merged**